usl_pipeline/usl_lib/tests/transformers/test_feature_raster_transformers.py

```diff
@@ -0,0 +1,117 @@
+    max_hydraulic_conductivity = max(
+        [item.hydraulic_conductivity for item in infiltration_configuration.items]
+    )
+    max_wetting_front_suction_head = max(
+        [item.wetting_front_suction_head for item in infiltration_configuration.items]
+    )
+    max_effective_porosity = max(
+        [item.effective_porosity for item in infiltration_configuration.items]
+    )
+    max_effective_saturation = max(
+        [item.effective_saturation for item in infiltration_configuration.items]
+    )
+
+    header = geo_data.ElevationHeader(
+        col_count=3,
+        row_count=4,
+        x_ll_corner=0.0,
+        y_ll_corner=0.0,
+        cell_size=1.0,
+        nodata_value=-9999.0,
+    )
+    data = numpy.asarray(
+        [
+            [0.0, 1.0, 2.0],
+            [3.0, 4.0, -9999.0],
+            [6.0, 7.0, -9999.0],
+            [9.0, 10.0, 11.0],
+        ]
+    )
+    elevation = geo_data.Elevation(header=header, data=data)
+    boundaries = [
+        (bbox_polygon(1, 0, 3, 4), 1),  # exclude the first column (row:col=*:0)
+    ]
+    buildings = [
+        (bbox_polygon(0, 3, 3, 4), 1),  # this is the first row (row:col=0:*)
+    ]
+    green_areas = [
+        (bbox_polygon(0, 0, 3, 2), 1),  # these are last two rows (row:col=2,3:*)
+    ]
+    soil_classes = [
+        (bbox_polygon(1, 1, 2, 2), 3),  # row:col=2:1
+        (bbox_polygon(1, 0, 2, 1), 4),  # row:col=3:1
+        (bbox_polygon(2, 0, 3, 1), 11),  # row:col=3:2
+    ]
+    feature_matrix = feature_raster_transformers.transform_to_feature_raster_layers(
+        elevation,
+        boundaries,
+        buildings,
+        green_areas,
+        soil_classes,
+        infiltration_configuration,
+    ).tolist()
+
+    # Checking the first column (should be NODATA since outside the boundaries)
+    assert feature_matrix[0][0] == [-9999.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
+    assert feature_matrix[1][0] == [-9999.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
+    assert feature_matrix[2][0] == [-9999.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
+    assert feature_matrix[3][0] == [-9999.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
+
+    # Checking the remaining first row (should be elevation/mask present + buildings=1)
+    assert feature_matrix[0][1] == [1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0]
+    assert feature_matrix[0][2] == [2.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0]
+
+    # Checking cells row:col={1:2},{2:2} (should be NODATA as in original elevation)
+    assert feature_matrix[1][2] == [-9999.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
+    assert feature_matrix[2][2] == [-9999.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
+
+    # Cell row:col=1:1 has just elevation (with mask=1), no buildings, no greens
+    assert feature_matrix[1][1] == [4.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
+
+    # Checking remaining 3 soil-related cells:
+    # Cell row:col=2:1, soil class=3
+    assert feature_matrix[2][1] == [
```

Comment:
`numpy.testing.assert_array_almost_equal` 
is helpful for testing arrays with floats. I'm actually surprised these ended up being exactly equal with floating point imprecision.
`assert_array_equal` and `assert_array_almost_equal` will give better error messages if things aren't equal. You can also pass them `strict=True` to ensure the two arrays have matching dtypes, which would be a good thing to assert here for float32s.

Reply:
Done